[WebXR] Only allow WebXR IPC messages after prompt is confirmed

mwyrzykowski opened the pull request:
#### 6afb7c249681f62c45b1566827e17a44731e9a97
<pre>
[WebXR] Only allow WebXR IPC messages after prompt is confirmed
<a href="https://bugs.webkit.org/show_bug.cgi?id=276380">https://bugs.webkit.org/show_bug.cgi?id=276380</a>
&lt;radar://128015279&gt;

Reviewed by Dan Glastonbury.

Make WebXR IPC calls no-ops if the prompt hasn&apos;t been accepted.

* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::immersiveModeProcessIdentity const):
(WebKit::GPUProcess::webXRPromptAccepted):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
(WebKit::RemoteGraphicsContextGL::webXRPromptAccepted):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(createExternalImage):
(deleteExternalImage):
(bindExternalImage):
(createExternalSync):
(addFoveation):
(enableFoveation):
(disableFoveation):
(framebufferDiscard):
(framebufferResolveRenderbuffer):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::webXRPromptAccepted):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::processIdentity):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::requestPermissionOnSessionFeatures):
(WebKit::PlatformXRSystem::initializeTrackingAndRendering):
(WebKit::PlatformXRSystem::shutDownTrackingAndRendering):
(WebKit::PlatformXRSystem::didCompleteShutdownTriggeredBySystem):
(WebKit::PlatformXRSystem::setImmersiveSessionState):
(WebKit::PlatformXRSystem::invalidateImmersiveSessionState):
* Source/WebKit/UIProcess/XR/PlatformXRSystem.h:

Canonical link: <a href="https://commits.webkit.org/281114@main">https://commits.webkit.org/281114@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/26d3ad9a33d409cfaa7696ea631ff3dff0e56496

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57827 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37155 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10303 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61449 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8272 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44791 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8460 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46851 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5872 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59857 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34847 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49975 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27679 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/57353 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31615 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7276 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53563 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7538 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63132 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1741 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7608 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54075 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1747 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49986 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54598 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13000 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1480 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32984 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34070 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35154 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33815 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
```diff
@@ -1641,13 +1641,19 @@
 #if ENABLE(WEBXR)
     void createExternalImage(uint32_t name, WebCore::GraphicsContextGL::ExternalImageSource&& arg0, uint32_t internalFormat, int32_t layer)
     {
+        if (!webXRPromptAccepted())
```
This needs to come from the `./Tools/Scripts/generate-gpup-webgl` script, I will modify that instead

djg replied:
Where is this change?

I believe you can update the prototypes in `RemoteGraphicsContextGLProxy.h` from:
```
    GCGLExternalImage createExternalImage(WebCore::GraphicsContextGL::ExternalImageSource&&, GCGLenum internalFormat, GCGLint layer) IPC_MESSAGE_ATTRIBUTE(EnabledIf='webXREnabled()') final;
```
to
```
    GCGLExternalImage createExternalImage(WebCore::GraphicsContextGL::ExternalImageSource&&, GCGLenum internalFormat, GCGLint layer) IPC_MESSAGE_ATTRIBUTE(EnabledIf='webXRPromptedAccepted()') final;
```
and the tooling will generate the check for you.

(Unless the checks need to be MESSAGE_CHECK)

mwyrzykowski replied:
The change was reverted after I updated `./Tools/Scripts/generate-gpup-webgl`

djg replied:
Why is the change to `generate-gpu-webgl` not part of this patch?

mwyrzykowski replied:
Oh forgot to update, the changes ended up being limited to RemoteGraphicsContextGL.messages.in https://github.com/WebKit/WebKit/pull/30612/files#r1672818808

djg commented on p:
```diff
@@ -0,0 +1,266 @@
+diff --git a/Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/ContextMtl.mm b/Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/ContextMtl.mm
```
Looks like you have a spurious addition

mwyrzykowski commented on Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
```diff
@@ -132,11 +132,14 @@ void PlatformXRSystem::requestPermissionOnSessionFeatures(const WebCore::Securit
-                protectedThis->setImmersiveSessionState(ImmersiveSessionState::PermissionsGranted);
-            } else
+                protectedThis->setImmersiveSessionState(ImmersiveSessionState::PermissionsGranted, [grantedFeatures = WTFMove(grantedFeatures), completionHandler = WTFMove(completionHandler)](bool) mutable {
+                    completionHandler(WTFMove(grantedFeatures));
```
Deleting the completionHandler(...) call on line 139 and adding it to after the async message replies should address the race, as far as I am aware

mwyrzykowski commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
```diff
@@ -323,19 +323,19 @@ messages -> RemoteGraphicsContextGL NotRefCounted Stream {
-    [EnabledIf='webXREnabled()'] void DeleteExternalImage(uint32_t handle)
-    [EnabledIf='webXREnabled()'] void BindExternalImage(uint32_t target, uint32_t arg1)
-    [EnabledIf='webXREnabled()'] void CreateExternalSync(uint32_t externalSync, WebCore::GraphicsContextGL::ExternalSyncSource arg0) NotStreamEncodable
+    [EnabledIf='webXRPromptAccepted()'] void CreateExternalImage(uint32_t externalImage, WebCore::GraphicsContextGL::ExternalImageSource arg0, uint32_t internalFormat, int32_t layer) NotStreamEncodable
```
these are the changes needed so we don't need to modify the generate-gpuwebgl script

djg replied:
You need to make the changes to `RemoteGraphicsContextGLProxy.h` as I pointed out in https://github.com/WebKit/WebKit/pull/30612#discussion_r1671648125. Then you run `generate-gpup-webgl` script to update the automatically generated message handling. If you don't update `RemoteGraphicsContextGLProxy.h`, next time someone runs the script, the changes will be undone.

mwyrzykowski replied:
Oh thank you, makes sense, I’ll change that. That is probably what happened to my changes last time

mwyrzykowski commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
```diff
@@ -361,20 +361,20 @@ class RemoteGraphicsContextGLProxy : private IPC::Connection::Client, public Web
-    void deleteExternalImage(GCGLExternalImage handle) IPC_MESSAGE_ATTRIBUTE(EnabledIf='webXREnabled()') final;
-    void bindExternalImage(GCGLenum target, GCGLExternalImage) IPC_MESSAGE_ATTRIBUTE(EnabledIf='webXREnabled()') final;
-    GCGLExternalSync createExternalSync(WebCore::GraphicsContextGL::ExternalSyncSource&&) IPC_MESSAGE_ATTRIBUTE(EnabledIf='webXREnabled()') final;
+    GCGLExternalImage createExternalImage(WebCore::GraphicsContextGL::ExternalImageSource&&, GCGLenum internalFormat, GCGLint layer) IPC_MESSAGE_ATTRIBUTE(EnabledIf='webXRPromptAccepted()') final;
```
Like this right @djg?

djg reviewed (approved):
:shipit:

mwyrzykowski committed:
[WebXR] Only allow WebXR IPC messages after prompt is confirmed
https://bugs.webkit.org/show_bug.cgi?id=276380
<radar://128015279>

Reviewed by Dan Glastonbury.

Make WebXR IPC calls no-ops if the prompt hasn't been accepted.

* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::immersiveModeProcessIdentity const):
(WebKit::GPUProcess::webXRPromptAccepted):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
(WebKit::RemoteGraphicsContextGL::webXRPromptAccepted):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(createExternalImage):
(deleteExternalImage):
(bindExternalImage):
(createExternalSync):
(addFoveation):
(enableFoveation):
(disableFoveation):
(framebufferDiscard):
(framebufferResolveRenderbuffer):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::webXRPromptAccepted):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::processIdentity):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::requestPermissionOnSessionFeatures):
(WebKit::PlatformXRSystem::initializeTrackingAndRendering):
(WebKit::PlatformXRSystem::shutDownTrackingAndRendering):
(WebKit::PlatformXRSystem::didCompleteShutdownTriggeredBySystem):
(WebKit::PlatformXRSystem::setImmersiveSessionState):
(WebKit::PlatformXRSystem::invalidateImmersiveSessionState):
* Source/WebKit/UIProcess/XR/PlatformXRSystem.h:

Canonical link: https://commits.webkit.org/281114@main

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -261,6 +261,16 @@ bool RemoteGraphicsContextGL::webXREnabled() const
+bool RemoteGraphicsContextGL::webXRPromptAccepted() const
+{
+#if ENABLE(WEBXR) && PLATFORM(COCOA) && !PLATFORM(IOS_FAMILY_SIMULATOR)
+    auto currentAcceptedValue = GPUProcess::singleton().immersiveModeProcessIdentity();
```
This is not thread safe

mwyrzykowski replied:
I think with https://github.com/WebKit/WebKit/pull/31404 the lack of thread safety in this member function is a non-issue, since `webXRPromptAccepted` is only called during immersive mode or shortly after the immersion session has ended. It would not be possible from the UI layer to reproduce a race here in that window.

kkinnunen-apple replied:
It's definitively possible to se torn read. GL reads this in GL work queue, the message handler for the UI process -sent message write this in main thread.

kkinnunen-apple reviewed (commented):
This is not correct.
For example, there's nothing guaranteeing that commands being issued for immersive session in the WCP -> GPUP are executed when the UI stops the immersive mode with UI -> GPUP. This will mean that commands will be skipped / reported as invalid, and the process will error out / be terminated.
